### ISSUE TITLE
Window size and ct_size setting when resolution_h doesn't equal resolution_w

### ISSUE THREAD

**Issue by QingXIA233** [NVlabs/FasterViT#3]
Hello, really impressive work! I notice that Faster-vit is trained on ImageNet with resolution (224, 224). In the code, everything is writen assuming that resolution_h equals reolution_w. However, if I need to use Faster-ViT (eg. faster_vit_0) for other resolutions such as (576, 960), here the size isn't 224 and resolution_h doesn't equal resolution_w, how should I set the window size and ct_size so that the model could achieve its best performance? Really need some advice here, thanks in advance.

**Comment by ahatamiz** [comment]
Hi @QingXIA233 , thanks for the note. I will add any-resolution support in a couple of days. 

The `ct_size` is a hyper-parameter with performance vs latency tradeoff. Sampling larger carrier tokens definitely improves the performance but with a small hit on latency. In the supplementary materials (Table S3), we have ablation studies regarding this point. Increasing `window_size` on the other hand would definitely have a higher impact on latency (and the reason why we started thinking about HAT). Our idea is to keep the `window_size` as small as possible while letting the carrier tokens cross-communicate across regions. 

We believe HAT is an scalable, and latency-friendly, way of computing attention especially high-resolutions. 

Please stay tuned for the new addition to the code for any-resolution support.

**Comment by QingXIA233** [comment]
Thank you for your response at such short notice. I'm really looking forward to the resolution-related code enhancement. I would like to close this issue when it's done. Hope it's ok for you. Thanks.

**Comment by ahatamiz** [comment]
Hi @QingXIA233 , I have pushed the [any-resolution FasterViT](https://github.com/NVlabs/FasterViT/blob/main/models/faster_vit_any_res.py) model that should address this feature request. I also added logic for accommodating odd resolution and window sizes by properly padding the input features. 

I tested the following model which seems to work nicely:

```
 def faster_vit_0_any_res(pretrained=False, **kwargs):
    drop_path_rate = kwargs.pop("drop_path_rate", 0.2)
    resolution = kwargs.pop("resolution", [576, 960])
    model = FasterViT(depths=[2, 3, 6, 5],
                      num_heads=[2, 4, 8, 16],
                      window_size=[7, 7, 12, 12],
                      ct_size=2,
                      dim=64,
                      in_dim=64,
                      mlp_ratio=4,
                      resolution=resolution,
                      drop_path_rate=drop_path_rate,
                      **kwargs)
    model.default_cfg = default_cfgs['faster_vit_0_any_res']
    if pretrained:
        model.load_state_dict(torch.load(pretrained))
    return model
```

Regarding your question on `ct_size` vs `window_size`, the trade-off should be studied for each applications separately. For segmentation, detection, one may be able to use extremely small window sizes (e.g. 6 or 4) and keeping `ct_size` to a reasonable value. Needless to mention, by increasing the window size, the `ct_size` can be gracefully decreased --although having large carrier tokens should still not hurt. 

From a high-level understanding, carrier tokens should be large enough to properly capture the semantics of each local region. For ImageNet experiments, we deal with local regions of 7 x 7. Hence, 2 x 2 carrier token makes sense. As the size of local window sizes grow, the `ct_size` should be adjusted properly -- running ablation study on this parameter is highly recommended.

Thanks again for the great suggestion in adding the any-resolution FasterViT !

**Comment by QingXIA233** [comment]
Hello, @ahatamiz , I'm really impressed by your promptness of soving the problem. Yes, I used your any-resolution Faster-ViT model as the backbone of my 3D detection model. Everything works well with resolution (576, 960) ! As a matter of fact, I even converted the model to TensorRT to get the latency comparison. When I use Next-ViT(tiny version) as backbone, the latency is around 46ms; After changing the backbone to faster_vit_0, the latency is 50ms, which is still in the acceptable range. Therefore, the model with faster-vit-0 as backbone is currently being trained to see whether it can achieve better performance than the one with Next-ViT. Fingers crossed! Thank you for your help and the useful information. Close the issue now.

**Comment by ahatamiz** [comment]
Thanks you @QingXIA233  for the note. I am excited to know about the results in the 3D detection experiments. 

The fastervit-0 model latency can be still decreased by attaining a lower `window_size` or lowering the `ct_size` in my opinion. However, this could impact the accuracy. The application of carrier tokens in downstream tasks still remain an interesting subject to delve into. 

Best Regards

**Comment by pianogGG** [comment]
Hi @ahatamiz and @QingXIA233 , https://github.com/NVlabs/FasterViT/blob/main/fastervit/models/faster_vit_any_res.py#L666C28-L666C43 from here, we can see seq_length=cr_tokens_total, and cr_tokens_total = cr_tokens_per_window*sr_ratio[0]*sr_ratio[1] , and https://github.com/NVlabs/FasterViT/blob/main/fastervit/models/faster_vit_any_res.py#L271  relative_bias = torch.zeros(1, num_heads, seq_length, seq_length), so how can we implement the loading of pretrained models on any [h,w] to achieve dynamic [h,w] size support? Thanks a lot
